### ISSUE TITLE
Support `output` in batch level methods of callaback

### ISSUE THREAD

**Issue by innat** [keras-team/tf-keras#330]
**System information**.

TensorFlow version (you are using): 2.9
Are you willing to contribute it (Yes/No) : No (It's not clear from where to start; it may also required API desgin level discussion.)

## Describe the feature and the current behavior/state

Currently, if we want to use a data set in callback, we usually pass it to the corresponding callback's constructor. That dataset can be validation set or any special dataset that we want to `.evaluate` or `.predict`, etc. The feature of current callback methods are listed [here](https://keras.io/guides/writing_your_own_callbacks/).

Now, I've encountered a case, where I need to have the following from validation set.

- `image_path`, 
- `image_tensor`,
-  `target`, 
- `predicted` 
- '<>`

I've a `.fit` method as follows, and I need to use a callback for validation (`on_test_batch_end`) data to further save the image-tensor as image-file with predicted value (here, which is mask) and various type of metrics calculation at instance level (samplw-wise). 

```python
model.fit(
    training_dataset, 
    validation_data=valid_dataset, 
    epochs=EPOCHS,
)
```

- I am not sure if **epoch-level** or **global-level** need such modificaiton. 
- We can of course, run the model on the validation dataset inside the callback and get true and predicted results; but IMHO, that might be a wrong argument. By this, we're running the validation twice, one for `validation_data` in `.fit` and another time in callback, looks unncessary. Also, there can be multiple callback which may require the same.
- There can be some hacks to achive this but it might be more natural to have those attributes (`image_paths`, `tensor`, `target` and `prediction`, [and so on] ) from callback-level method.
- PyTorch-Lightning [reference](https://pytorch-lightning.readthedocs.io/en/latest/_modules/pytorch_lightning/callbacks/callback.html#Callback.on_validation_batch_end).

**Will this change the current api? How?**

Currently
```python
on_(train|test|predict)_batch_end(self, batch, logs=None)
```
Maybe, 
```python
on_(train|test|predict)_batch_end(self, output, batch_id, logs=None)

outputs["image_path"],
outputs["image"],
outputs["gt_label"],
outputs["prediction"],
```

Here the term **output** ( also in the post title) inspired from pytorch-lightning and just for demonstration purpose).

**Who will benefit from this feature?**

keras-user.

**[Contributing](https://github.com/keras-team/keras/blob/master/CONTRIBUTING.md)**

- Do you want to contribute a PR? (yes/no): no
- If yes, please read [this page](https://github.com/keras-team/keras/blob/master/CONTRIBUTING.md) for instructions
- Briefly describe your candidate solution(if contributing):

**Comment by innat** [comment]
**summarized workaround**
https://stackoverflow.com/a/76674353/9215780

**Comment by innat** [comment]
@fchollet 
Same here either. Wasn't it fit keras? Why move here? This issue doesn't related to tf-keras but keras.
